### DISCUSSION TITLE
Is it possible to schedule a post in Quarto?

### DISCUSSION THREAD

**Discussion by maikol-solis** [quarto-dev/quarto-cli#3114] [Q&A]
Hi! 

I have set it up a quarto blog in github, but I am wondering if it's possible to schedule my posts to some specific day in the future?

I appreciate any advise or workflow in the matter. 

Thanks

**Comment by dragonstyle** [comment]
I'm not sure about how Github would support scheduled rendering of documents, but one thing I can point to being a piece of the puzzle is the `draft` option for documents - if you set `draft: true`, the document will be not be included in listings (blogs), so you could write a document and mark it as a draft and have some scheduled task change the draft to false, then re-render it...

**Reply by cscheid** [reply]
I wonder if we could write a filter that sets `draft: true` to the metadata any posts with dates in the future of the rendering date. That would do it, right?

**Reply by mcanouil** [reply]
This would be great, especially with a GHA triggered using CRON.

**Reply by maikol-solis** [reply]
> I wonder if we could write a filter that sets `draft: true` to the metadata any posts with dates in the future of the rendering date. That would do it, right?

This would be awesome. 

Hugo has a similar feature 

https://gohugo.io/getting-started/configuration/#configure-dates

**Reply by cscheid** [reply]
If anyone wants to write that filter, we'd love to feature it in `quarto-examples`. It should be a matter of reading the system date while handling document metadata in a filter, something like:

```lua
Meta = function(meta)
  if is_after(meta.date, today()) then -- Note that I don't know how lua date functions work, and it very likely won't be "today()"!
    meta["draft"] = true
    return meta
  end
end
```

**Reply by dragonstyle** [reply]
It does raise the question of whether listings should exclude pages with dates in the future by default (or at least have a mode to do this).

**Reply by mcanouil** [reply]
It's not only about pages or child documents for listings, especially since listings can be built using only yaml.

**Reply by cscheid** [reply]
> It's not only about pages or child documents for listings, especially since listings can be built using only yaml.

@mcanouil I don't understand the point you're trying to make. Can you clarify?

**Reply by mcanouil** [reply]
If the same kind of mechanism future date = draft is to be applied for listings, this implies all kind of listings, not only the ones based on child Quarto documents.
For example in my work in progress website, I am using the following listings, which are not all using "date" but they could:
- https://github.com/mcanouil/mickael.canouil.fr/blob/fix-issue23/projects.qmd
- https://github.com/mcanouil/mickael.canouil.fr/blob/fix-issue23/publications.qmd
- https://github.com/mcanouil/mickael.canouil.fr/blob/fix-issue23/talks.qmd

I have to admit that I do not know how Quarto is dealing with those listings formats.

**Comment by mcanouil** [comment]
You could set a CRON task in GitHub Actions and setup some workflow/steps to remove any draft or simply render the website/document.

```
on:
  workflow_dispatch:
  schedule:
    - cron: "0 12 * * mon,wed,fri"
```

**Reply by jimjam-slam** [reply]
If the blog is on GitHub Pages (or another JAMstack provider that triggers a render when a branch is committed to, like [Netlify](https://netlify.com)), it could even be as simple as writing the blog post on a branch, creating a PR for it and then scheduling the PR merge! Using, for example: https://github.com/marketplace/actions/merge-schedule

**Reply by maikol-solis** [reply]
This is an excellent workaround. Thanks!

**Comment by andrewpbray** [comment]
I tried my hand at implementing the filter described by @cscheid . It can be found here:

[https://github.com/andrewpbray/scheduled-posts](https://github.com/andrewpbray/scheduled-posts)

It seems to work at appending `draft: true` to the meta data for documents with dates in the future, however I haven't gotten that to translate into the listing actually not showing those rendered documents. A repo demonstrating this behavior can be found here:

[https://github.com/andrewpbray/scheduled-posts-example/tree/1ec797df398934079c3d6bc47b7c4d7129616222](https://github.com/andrewpbray/scheduled-posts-example/tree/1ec797df398934079c3d6bc47b7c4d7129616222)

Any ideas on how to troubleshoot this? My go-to solution has been rendering to native format and inspecting the AST, but `quarto render posts.qmd --to native` doesn't show any of the content of the listing and the `listing.json` does show all three posts. Does it have something to do with the order in which various filters are being run?

ps: the way this filter does the date comparison, using `date-format: "X"` seems like a bad idea and something to improve on.

**Reply by cscheid** [reply]
Hm. I think this means my approach won't work 😞. What's happening is that the listings functionality reads the metadata from the _input_, and not from the _output_, and the we don't realize that filters could have added metadata. This is a big bummer.

Let me talk with @dragonstyle and we'll get back to you.

**Reply by andrewpbray** [reply]
Ah, that makes sense. Thanks!

As I've been tinkering with this, I realize even when this works it will only fit about 3/4 of my use case. The functionality around `draft: true` is very narrowly about listings and search. I've realized that what I want is something broader: I want to render .qmds subject to the `date` metadata in a doc.

My plan for this is to write a pre-render script that copies all of the docs not in the future into a directory that is subject to a targeted render, then a post-render script that cleans out that directory. Ideally this could be accomplished script-free in the `_quarto.yml` directly, but . . . I imagine this functionality won't come along anytime soon.

**Comment by andrewpbray** [comment]
@cscheid Here's an approach that does seem to work:

[https://github.com/andrewpbray/scheduled-posts-example](https://github.com/andrewpbray/scheduled-posts-example)

It uses a pre-render script that edits the `QUARTO_PROJECT_INPUT_FILES` variable by removing files that are in the future.

It's not as elegant as calling a filter, but it has the advantage of making the future posts completely absent from the website (instead of just unlinked).

**Comment by andrewpbray** [comment]
@maikol-solis I think I've finally settled upon a robust way to do this.

[https://github.com/qmd-lab/scheduled-docs](https://github.com/qmd-lab/scheduled-docs)

**Reply by maikol-solis** [reply]
Thanks! I'll check this out.
